src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -728,44 +735,54 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded, Immu
                     // TODO: allow finalize compilation to incrementally update a prior version
                     // https://github.com/dotnet/roslyn/issues/46418
                     var compilationWithoutGeneratedFiles = compilation;
-                    var generatedDocumentsBuilder = ImmutableArray.CreateBuilder<SourceGeneratedDocumentState>();
 
-                    if (generators.Any())
-                    {
-                        var additionalTexts = this.ProjectState.AdditionalDocumentStates.Values.SelectAsArray(a => (AdditionalText)new AdditionalTextWithState(a));
-                        var compilationFactory = this.ProjectState.LanguageServices.GetRequiredService<ICompilationFactoryService>();
+                    ImmutableArray<SourceGeneratedDocumentState> generatedDocuments;
 
-                        var generatorDriver = compilationFactory.CreateGeneratorDriver(
-                                this.ProjectState.ParseOptions!,
-                                generators,
-                                this.ProjectState.AnalyzerOptions.AnalyzerConfigOptionsProvider,
-                                additionalTexts);
+                    if (authoritativeGeneratedDocuments != null)
+                    {
+                        generatedDocuments = authoritativeGeneratedDocuments.Value;
+                    }
+                    else
+                    {
+                        var generatedDocumentsBuilder = ImmutableArray.CreateBuilder<SourceGeneratedDocumentState>();
 
-                        if (generatorDriver != null)
+                        if (generators.Any())
                         {
-                            generatorDriver = generatorDriver.RunGenerators(compilationWithoutGeneratedFiles);
+                            var additionalTexts = this.ProjectState.AdditionalDocumentStates.Values.SelectAsArray(a => (AdditionalText)new AdditionalTextWithState(a));
```

Comment:
there's a CastUp helper for this.

Reply:
Nifty, since it doesn't allocate.

Reply:
Oh it's not an extension method though, boo. Going to leave this as is.